test/peer-discovery/index.spec.js

```diff
@@ -10,57 +10,108 @@ const defer = require('p-defer')
+    let libp2p
 
-  before(async () => {
-    [peerInfo, remotePeerInfo] = await createPeerInfo({ number: 2 })
-  })
-
-  afterEach(async () => {
-    libp2p && await libp2p.stop()
-    sinon.reset()
-  })
+    before(async () => {
+      [peerInfo, remotePeerInfo] = await createPeerInfo({ number: 2 })
+    })
 
-  it('should dial know peers on startup', async () => {
-    libp2p = new Libp2p({
-      ...baseOptions,
-      peerInfo
+    afterEach(async () => {
+      libp2p && await libp2p.stop()
+      sinon.reset()
     })
-    libp2p.peerStore.add(remotePeerInfo)
-    const deferred = defer()
-    sinon.stub(libp2p.dialer, 'connectToPeer').callsFake((remotePeerInfo) => {
-      expect(remotePeerInfo).to.equal(remotePeerInfo)
-      deferred.resolve()
+
+    it('should dial know peers on startup', async () => {
+      libp2p = new Libp2p({
+        ...baseOptions,
+        peerInfo
+      })
+      libp2p.peerStore.add(remotePeerInfo)
+      const deferred = defer()
+      sinon.stub(libp2p.dialer, 'connectToPeer').callsFake((remotePeerInfo) => {
+        expect(remotePeerInfo).to.equal(remotePeerInfo)
+        deferred.resolve()
+      })
+      const spy = sinon.spy()
+      libp2p.on('peer:discovery', spy)
+
+      libp2p.start()
+      await deferred.promise
+      expect(spy.getCall(0).args).to.eql([remotePeerInfo])
     })
-    const spy = sinon.spy()
-    libp2p.on('peer:discovery', spy)
 
-    libp2p.start()
-    await deferred.promise
-    expect(spy.getCall(0).args).to.eql([remotePeerInfo])
+    it('should ignore self on discovery', async () => {
+      libp2p = new Libp2p(mergeOptions(baseOptions, {
+        peerInfo,
+        modules: {
+          peerDiscovery: [MulticastDNS]
+        }
+      }))
+
+      await libp2p.start()
+      const discoverySpy = sinon.spy()
+      libp2p.on('peer:discovery', discoverySpy)
+      libp2p._discovery[0].emit('peer', libp2p.peerInfo)
+
+      expect(discoverySpy.called).to.eql(false)
+    })
   })
 
-  it('should ignore self on discovery', async () => {
-    libp2p = new Libp2p(mergeOptions(baseOptions, {
-      peerInfo,
-      modules: {
-        peerDiscovery: [MulticastDNS]
-      }
-    }))
+  describe('discovery modules from transports', () => {
+    let peerInfo, libp2p
 
-    await libp2p.start()
-    const discoverySpy = sinon.spy()
-    libp2p.on('peer:discovery', discoverySpy)
-    libp2p._discovery[0].emit('peer', libp2p.peerInfo)
+    before(async () => {
+      [peerInfo] = await createPeerInfo()
+    })
+
+    it('should add discovery module if present in transports and enabled', async () => {
+      libp2p = new Libp2p(mergeOptions(baseOptions, {
+        peerInfo,
+        modules: {
+          transport: [WebRTCStar]
+        },
+        config: {
+          peerDiscovery: {
+            webRTCStar: {
+              enabled: true
+            }
+          }
+        }
+      }))
+
+      await libp2p.start()
+
+      expect(libp2p._discovery).to.have.lengthOf(1)
```

Comment:
We should verify this is the discovery service we expect. Making `libp2p. _discovery` a map would make this easy to do.